### ISSUE TITLE
Add a SCons option to compile modules as shared libraries for faster rebuilds

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#1796]
Related to https://github.com/godotengine/godot/issues/41271 and https://github.com/godotengine/godot/issues/42048.

**Describe the project you are working on:**

The Godot editor :slightly_smiling_face: 

**Describe the problem or limitation you are having in your project:**

Even on a high-end CPU, performing an incremental build of Godot can take a long time compared to other codebases of a similar size. This is in part due to slow linking times. Using `use_llvm=yes use_lld=yes` helps on Linux, but it's still relatively slow.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

Add a `linking` SCons option which accepts the `static` (default) or `shared` value. The current behavior is `static`, so we'd have to implement the required logic in SCons to build shared libraries when using `linking=shared`.

Since the libraries will be smaller and relinked only when needed, this allows the linker to do less work to provide an incremental build of Godot. This could also help speed up core builds since the core binary will be smaller.

Using shared libraries will likely only be supported for development purposes, not in release builds.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

We have some [documentation](https://docs.godotengine.org/en/latest/development/cpp/custom_modules_in_cpp.html#improving-the-build-system-for-development) on using dynamic linking for third-party modules, but we still have to figure out how to apply this on a more global scale. For example, this could be applied to all built-in modules.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

No, as this is core buildsystem functionality.

**Is there a reason why this should be core and not an add-on in the asset library?:**

This is core buildsystem functionality.

**Comment by jasonliu--** [comment]
Would this proposal allow for the potential to link against externally-compiled shared libraries?

In MacPorts (and all *nix OS package managers), we have all of the libraries in the `thirdparty` folder already compiled as individual packages. In addition, nearly all of those libraries are shared with Blender, which I have packaged to (dynamically) link against the packaged libraries, instead of allowing Blender to compile the libraries internally as part of Blender's build process (regardless of whether they would be statically or dynamically linked).

If your proposal would allow for the possibility of linking against externally-built libs, then not only would it allow for shorter build times because it would no longer be Godot's responsibility to compile them, but if you also allow for dynamic linking (which, as you have already pointed out, is already somewhat possible), it would also allow for less memory usage when Blender and Godot are both running on the same machine, which I suspect is a very common situation among people who are using Godot for gamedev.

**Comment by Calinou** [comment]
> Would this proposal allow for the potential to link against externally-compiled shared libraries?

It's already feasible to *unbundle* non-modified thirdparty libraries (as is done by Linux distribution packages), but this has only been done on Linux so far. Remember that Godot uses many modified third-party libraries, and these can't be unbundled.

**Comment by jasonliu--** [comment]
I tried unbundling the libraries in MacPorts, and let's just say that "it didn't go well". 😭 Since you say that it's only been done on Linux so far, then that explains my experience. 😆 

> Remember that Godot uses many modified third-party libraries, and these can't be unbundled.

Then maybe, as a part of this proposal's efforts, it would be worthwhile to take a closer look at what sorts of modifications Godot is making to those third-party libraries, and to see whether there might be any way to use the stock, unmodified library, and move the "augmentations" in-house, over to Godot's side of the `#include` statement. I'm always nervous when making this sort of suggestion, because all programmers (including me) always feel that we're only modifying someone else's code because it's "absolutely necessary", and that there would be no other way to get the code to do what I want it to do. I'm nervous because I certainly don't want to insult the fine folks working on Godot's source, but sometimes we have to take a look in the mirror and ask, "is modifying the upstream code really the _only_ way to accomplish what I need, or am I just doing it because it's the _easiest_ way to accomplish what I need?"

In addition, if the modifications to the third-party library really truly are necessary, then perhaps it's possible that the Godot devs have uncovered a weakness in the library. The weakness might turn out to be some sort of bug, or even just something that the library "should" or "could" be doing, but hasn't implemented yet. In that case, maybe it would be worth the time and effort to contribute the modifications upstream, so that they get incorporated into the library, even if it gets implemented as an advanced option that only Godot ever uses. I realize that these sorts of things can sometimes take a long time, but if you're able to do that, then you are back to being able to link against a stock, unmodified version of the library.

If you are able to link against the unmodified library, then it means that it's one more piece that you can unbundle, and thus you can offload one more chunk of the compile time. I'm sure that the core dev team has already thought of this many times, but sometimes it helps to take a step back and to be ruthless to ourselves, and say "ok, we're going _force_ ourselves to find a way to unbundle this third-party library by leaving it unmodified, how do we accomplish that?" My guess is that if you are able to do that with (nearly) all of your third-party libraries, you will see some fairly markedly improved build times.

**Comment by bruvzg** [comment]
> Since you say that it's only been done on Linux so far, then that explains my experience. 😆

To replace built-in libraries, you can add `builtin_{library name}=no` to the build command, and manually set `CCFLAGS` and `LINKFLAGS` to have paths to your libs.

Also, most of the Linux `pkg-config` library detection code can be re-used on macOS without major changes. I have not tested it with MacPorts, but it's working fine with the Homebrew.

Patch for `pkg-config`: https://gist.github.com/bruvzg/ae78ae2fe05868e00024edc1ad61168d

```
export PKG_CONFIG_PATH=/opt/homebrew/opt/icu4c/lib/pkgconfig:/opt/homebrew/opt/zlib/lib/pkgconfig

scons tools=yes target=debug builtin_freetype=no builtin_graphite=no builtin_icu=no builtin_harfbuzz=no builtin_libpng=no builtin_bullet=no builtin_libtheora=no builtin_libvorbis=no builtin_libogg=no builtin_libwebp=no builtin_wslay=no builtin_miniupnpc=no builtin_pcre2=no builtin_embree=no builtin_zlib=no warnings=extra werror=yes use_volk=yes
```

**Comment by jasonliu--** [comment]
@bruvzg that's quite good, it looks like you copied that from `platform/x11/detect.py`? And you say that it's working on Homebrew? I'll give that a try on MacPorts, thanks for the tip! 😄

**Comment by bruvzg** [comment]
>  that's quite good, it looks like you copied that from `platform/x11/detect.py`

Yes, it's a copy from `platform/x11/detect.py` with the X11 stuff removed and a few hard-coded paths changed to `/opt/homebrew/` (MacPorts and Homebrew on Intel macs will need another path, but it should be fine to add all variants at the same time).

**Comment by jasonliu--** [comment]
> > that's quite good, it looks like you copied that from `platform/x11/detect.py`
> 
> Yes, it's a copy from `platform/x11/detect.py` with the X11 stuff removed and a few hard-coded paths changed to `/opt/homebrew/` (MacPorts and Homebrew on Intel macs will need another path, but it should be fine to add all variants at the same time).

It works! Since @bruvzg seems to have it working on Homebrew, and I have gotten it working on MacPorts, is this something that I could submit a PR for, to get it incorporated into the codebase?
